csaf-matching/src/commonMain/kotlin/io/github/csaf/sbom/matching/Matcher.kt

```diff
@@ -16,16 +16,99 @@
- * specific criteria.
+ * Matcher for matching SBOM documents with provided CSAF documents.
  *
- * @param sbom The SBOM document represented by a protobom.protobom.Document instance.
- * @param doc The CSAF document to be matched against, represented by a Csaf instance.
- * @return A boolean value indicating whether the SBOM document matches the CSAF document.
+ * @property docs The CSAF documents used by this matcher.
+ * @property threshold The default threshold required for a match to be included.
  */
-fun match(sbom: protobom.protobom.Document, doc: Csaf): Boolean {
-    return true
+class Matcher(val docs: List<Csaf>, val threshold: Float = 0.5f) {
+    val purlMap = mutableMapOf<String, MutableSet<FastHash<Csaf>>>()
+    val cpeMap = mutableMapOf<Cpe, MutableSet<FastHash<Csaf>>>()
+
+    /**
+     * The constructor checks that the given threshold is within its bounds and then extracts all
+     * relevant information for comparison from the given CSAF documents for faster matching.
+     */
+    init {
+        require(threshold in 0.0..1.0) { "Threshold must be in the interval [0.0; 1.0]." }
+        docs.forEach { doc ->
+            // Cache all canonicalized PURLs
+            doc.product_tree
+                .mapBranchesNotNull {
+                    it.product?.product_identification_helper?.purl?.let {
+                        Purl(it.toString()).canonicalize()
+                    }
+                }
+                .forEach {
+                    purlMap.computeIfAbsent(it) { hashSetOf<FastHash<Csaf>>() }.add(FastHash(doc))
+                }
+            // Cache all CPEs, start with full product names
+            doc.product_tree?.full_product_names?.forEach {
+                it.product_identification_helper?.cpe?.let {
+                    cpeMap
+                        .computeIfAbsent(parseCpe(it)) { hashSetOf<FastHash<Csaf>>() }
+                        .add(FastHash(doc))
+                }
+            }
+            // Collect more CPEs from product tree (branches)
+            doc.product_tree
+                .mapBranchesNotNull {
+                    it.product?.product_identification_helper?.cpe?.let { parseCpe(it) }
+                }
+                .forEach {
+                    cpeMap.computeIfAbsent(it) { hashSetOf<FastHash<Csaf>>() }.add(FastHash(doc))
+                }
+        }
+    }
+
+    /**
+     * Matches the provided SBOM document with the CSAF documents and determines whether they meet
+     * specific criteria.
+     *
+     * @param sbom The SBOM document represented by a protobom.protobom.Document instance.
+     * @param threshold The minimum threshold required for a match to be included, defaults to the
+     *   value of this [Matcher].
+     * @return A list of CSAF documents matching the given SBOM, along with resp. match scores.
+     */
+    fun match(sbom: Document, threshold: Float = this.threshold): List<Match> {
```

Comment:
Please split this into a `matchAll` which works on a `Document` and a `match` which works on a single `Node`, since we only get a single node per matcher in DependencyTrack.

Reply:
done